### ISSUE TITLE
encoding of ruby glosses

### ISSUE THREAD

**Issue by martinascholger** [TEIC/TEI#2054]
This issue functions as a central hub for the work on the encoding of ruby glosses in TEI proposed by the East Asian / Japanese SIG. This includes the original proposal ([https://docs.google.com/document/d/1rRgBRYpxUlsnMEHbOPW-92u_AIMNwq9axGtpI1UGgz8/edit#](https://docs.google.com/document/d/1rRgBRYpxUlsnMEHbOPW-92u_AIMNwq9axGtpI1UGgz8/edit#)), the draft of the ODD, and the work on different examples. The proposal has been discussed together with @knagasaki, @nakamura196, and @kzhr as well as @martindholmes during the Council's virtual F2F meeting on October 18.

The ruby gloss, or _rubi_, or _furigana_, is a device for giving guidance on phonation, as well as presenting another reading for the text, which sometimes can be an essential part of the text. The structure of ruby glosses has, until recently, been established between characters and glosses, rather than words and glosses, since over a millennium ago. Such a textual structure requires another layer of semantics besides the existing set of vocabulary, such as `<seg>`, `<note>` (or `<add>`) and `<span>`. 
Therefore, the East Asian / Japanese SIG proposes to add `<ruby>`, `<rb>`, and `<rt>` to the TEI Guidelines to encode ruby glosses. 

The main points of discussion (in addition to the proposal): 

- to indicate the position of the glosses, the `@place` attribute will be used instead of `@rend`. The suggested values 'right' and 'left' are useful for both, the present proposal and for vertical aligned text in general.

- consider the values `default` and `additional` for the `@type` attribute instead of `primary` and `secondary`.

- method of fragmentation of the ruby text (e.g. kanji) to align the ruby reading to it (e.g. hiragana). There will likely be two forms, one with nesting structure and one relying on standoff in order to cover the most complex cases.

-  suggestion to show the same example encoded in the various models being proposed.

-  integration of ruby glosses into the Guidelines, one possibility would be in the core module next to the `<note>` element. 

- expand the proposed content model to allow nested `<ruby>` elements.

**Comment by duncdrum** [comment]
No suprises i guess that i m heavily in favor of this. Some thoughts for discussion, to be fleshed out later:

For the Guidelines examples, i would suggest to not only focus on the highly detailed requirements outlined in the gdocs, but to also consider helper methods for when rubys appear:
-  in a highly predictable and regular manner (on every word or even on every character)
    In the first case a general instruction in the header should be sufficient to define how ruby text generally relates to its ruby base without proliferation of and xml:ids:
    ```xml
    <ruby>
      <rb></rb>
      <rt></rt>
    </ruby>
    ```

    in the main body. Editors should have a lighter option, or otherwise the extra work will lead to many ignoring ruby or cooking up custom solutions. E.g. `<layout>` seems a good place to collect general information about the appearance of ruby (left right) without repeating it on every ruby element. 

-  as hand written annotation (as in court documents).
    This is mainly a technical demand that `@hand` is available and all the usual manuscribal attributes and elements. 

The guidelines should:
-   offer a suggestion for how to encode missing sections in the ruby stream (`<gap>` doesn't strike me as the right choice here based on its description line). 

-  clarify how emphasis markers relate to ruby in TEI. Should emphasis markers be encoded as ruby, use as css via `@rend`, … what about texts with both emphasis markers and ruby?

-  demonstrate that `<gaiji>` can appear both in `<rb>` and `<rt>`

There is one more question about the multi-textuality that can be established by ruby text, but i need more time to flesh that out myself

**Comment by sydb** [comment]
First, here is a  [first crack at an ODD file](https://github.com/TEIC/TEI/files/5475933/ruby_schema.zip) for ruby annotations. That ZIP archive contains the generated RELAX NG, Schematron, and HTML, too.
Second, here is the
[XML file (with “.txt” extension to appease GitHub)](https://github.com/TEIC/TEI/files/5475941/ruby01.xml.txt) we were playing around with on our recent conference call. This file makes use of the `<r:sub>` element to segment annotations, an element which does not actually exist (yet?) in the ODD file or anywhere else.

**Comment by martindholmes** [comment]
I think we need one section in the Guidelines in core (CO) to cover the most straightforward cases, then a much more detailed explanation covering the more unusual cases raised by @duncdrum as well as standoff approaches in Characters, Glyphs and Writing Modes (WD); finally, I think there should be a mention in Linking, Segmentation and Alignment (SA) mentioning the standoff approach and pointing to the more detailed explanation in WD.

**Comment by martindholmes** [comment]
@duncdrum Could you post images of the more unusual cases you mention?

**Comment by sydb** [comment]
I think similarly to @martindholmes, but was imagining that the stand-off approach would be fully described in SA, with CO and WD directing the reader there if interested.

**Comment by duncdrum** [comment]
@martindholmes actually my concern was with long text in which non of the unusual cases occur, but yet i ll post images.

here is @sydb xml so its easier to track here:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="./ruby.rnc" type="application/relax-ng-compact-syntax"?>
<?xml-model href="./ruby.isosch" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<div xmlns="http://www.tei-c.org/ns/1.0"
   xmlns:r="http://www.tei-c.org/ns/proposal/ruby">
  <p>

    <r:ruby>
      <r:rb>
        <r:ruby>
          <r:rb>東<anchor/>南<anchor/></r:rb>
          <r:rt place="top" type="primary">とう<anchor/>なん<anchor/></r:rt>
        </r:ruby>
      </r:rb>
      <r:rt place="bottom" type="secondary">たつみ</r:rt>
    </r:ruby>

    <r:ruby>
      <r:rb>
        <r:ruby>
          <r:rb>
            <r:sub>東</r:sub>
            <r:sub>南</r:sub>
          </r:rb>
          <r:rt place="top" type="primary">
            <r:sub>とう</r:sub>
            <r:sub>なん</r:sub>
          </r:rt>
        </r:ruby>
      </r:rb>
      <r:rt place="bottom" type="secondary">たつみ</r:rt>
    </r:ruby>

  </p>
</div>

```

**Comment by martindholmes** [comment]
Doing some initial testing of @sydb's schema, I'm wondering if the ruby element should be allowed inside the w element. It seems to me that encoding linguistic segments in conjunction with ruby should be possible.

**Comment by martindholmes** [comment]
Posting a text document taking a very simple example and working out some implications. This uses a very useful single example from the Ruby team's proposal, which has the same three characters glossed in two different ways, with differing segmentation; the gloss on the left is a whole-word gloss of the whole kanji word, while the glosses on the right apply to only the first and second characters respectively.

![ruby_billiards](https://user-images.githubusercontent.com/2412617/104045091-e8ef3d00-5192-11eb-8a9d-30908133c0dc.png)

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!--<?xml-model href="ruby.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="ruby.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>-->
<TEI xmlns="http://www.tei-c.org/ns/1.0" xmlns:r="http://www.tei-c.org/ns/proposal/ruby">
  <teiHeader>
    <fileDesc>
      <titleStmt>
        <title/>
      </titleStmt>
      <publicationStmt>
        <p>Free to anyone for any purpose.</p>
      </publicationStmt>
      <sourceDesc>
        <p>Invented based on examples in prior work on this ticket.</p>
      </sourceDesc>
    </fileDesc>
  </teiHeader>
  <text>
    <body>
      <div>
        <p>All of these examples are based on a single very useful image of a real
        text which appears in the Ruby proposal.</p>

        <div>
          <head>Simple example</head>
          <p>Simple example taken from proposal document: billiard hall, with dual pronunciations
            <q>biriādo</q> and <q>dakyūba</q>. I've added the <q>ba</q> to the end of the right-side
            ruby text for simplicity, and used a supplied element to indicate that it's added.</p>
          <p>First, the simplest possible setup: the entire word is glossed in one:</p>
          <p style="writing-mode: vertical-rl"> 
            <r:ruby>
              <r:rb>打球場</r:rb>
              <r:rt place="left">ビリアード</r:rt>
              <r:rt place="right">ダキュウ<supplied>バ</supplied></r:rt>
            </r:ruby> 
          </p>
        </div>
        
        <div>
          <head>Character by character</head>
          <p>Next, we divide the characters with explicit ruby elements. This seems awkward; another approach
            would be to use pure contiguity to signal that every rt element applies to its preceding
            rb element, but contiguity has drawbacks too.</p>
          <p style="writing-mode: vertical-rl"> 
            <r:ruby>
              <r:rb>打</r:rb>
              <r:rt place="left">ビリ</r:rt>
              <r:rt place="right">ダ</r:rt>
            </r:ruby> 
            <r:ruby>
              <r:rb>球</r:rb>
              <r:rt place="left">アー</r:rt>
              <r:rt place="right">キュウ</r:rt>
            </r:ruby> 
            <r:ruby>
              <r:rb>場</r:rb>
              <r:rt place="left">ド</r:rt>
            </r:ruby> 
          </p>
        </div>
          
        <div>  
          <head>Alternative segmentation</head>
        <p>The actual example is intriguing in that the right-side ruby 
          glosses apply to the first and second characters respectively, but
          the left-side gloss applies to the whole word as a unit. We can 
          capture this using nested ruby elements.
        </p>
        <p style="writing-mode: vertical-rl">
          <r:ruby>
            <r:rb>
              <r:ruby>
                <r:rb>打</r:rb>
                <r:rt place="right">ダ</r:rt>
              </r:ruby> 
              <r:ruby>
                <r:rb>球</r:rb>
                <r:rt place="right">キュウ</r:rt>
              </r:ruby>
              場
            </r:rb>
            <r:rt place="left">ビリアード</r:rt>
          </r:ruby>
        </p>
        </div>
        
        <div>
          <head>Standoff ruby</head>
          <p>Next, we use anchors and pointers to map the relationships. This is effective, but it
          generates a lot of anchor elements.</p>
          <p style="writing-mode: vertical-rl"> 
            <r:ruby>
              <r:rb><anchor xml:id="da"/>打<anchor xml:id="kyuu"/>球<anchor xml:id="ba"/>場<anchor
                xml:id="owari"/></r:rb>
              <r:rt place="left" from="#da" to="#owari">ビリアード</r:rt>
              <r:rt place="right" from="#da" to="#kyuu">ダ</r:rt>
              <r:rt place="right" from="#kyuu" to="#ba">キュウ</r:rt>
            </r:ruby> 
          </p>
        </div>
        
        <div>
          <head>Another standoff approach using regular elements</head>
          <p>In place of anchors, we can use wrapping elements. This was not proposed by 
          the group, but @sydb and @martindholmes think it's a good approach:</p>
          <p style="writing-mode: vertical-rl"> 
            <r:ruby>
              <r:rb xml:id="dakyuuba"><c xml:id="da">打</c><c xml:id="kyuu">球</c>場</r:rb>
              <r:rt place="left" target="#dakyuuba">ビリアード</r:rt>
              <r:rt place="right" target="#da">ダ</r:rt>
              <r:rt place="right" target="#kyuu">キュウ</r:rt>
            </r:ruby> 
          </p>
        </div>
        
        <div>
          <head>Spans via anchors</head>
          <p>Another approach is to use anchors with no @xml:id to make parallel structures from which
            the relationships can be inferred. This will work as long as there is only one ruby text sequence, or
            where there are two or more, there are no internal differences in segmentation between them. 
            The second rt element simply ignores the final character.
          </p>
          
          <p style="writing-mode: vertical-rl"> 
            <r:ruby>
              <r:rb>打<anchor/>球<anchor/>場</r:rb>
              <r:rt place="left">ビリ<anchor/>アー<anchor/>ド</r:rt>
              <r:rt place="right">ダ<anchor/>キュウ<anchor/></r:rt>
            </r:ruby> 
          </p>
        </div>
      </div>
    </body>
  </text>
</TEI>

```

**Comment by sydb** [comment]
Is there any reason that a single `<ruby>` should be permitted to have multiple `<rb>` children? If so, does the number of `<rb>` children have to match the number of `<rt>` children?

**Comment by martindholmes** [comment]
@sydb I think (and our schema should now mandate) that a single `<ruby>` should allow only a single `<rb>`, followed by one or more `<rt>`s. Anything else is potentially very confusing.

**Comment by sydb** [comment]
OK. Here is the second [crack at an ODD file for ruby annotations](https://github.com/TEIC/TEI/files/5789805/ruby_schema.zip). That ZIP archive contains the generated RELAX NG, Schematron, and HTML, too.

This schema permits `<ruby>` inside `<w>` (alright, to be precise it permits `<r:ruby>`, but I am eliding the namespace prefix here), adds a `@target` to `<rt>`, and requires that _either_ `@target`, OR the combination of both `@from` & `@to`, OR none of these attributes are used on `<rt>`.

**Comment by martindholmes** [comment]
@sydb and I discussing the nature of the prose think that a good way to introduce the whole issue (thinking of readers not necessarily familiar with the concept) would be to follow the W3C's lead in starting with the example of the organization's initialism (TEI) with Ruby showing its expansion.

**Comment by sydb** [comment]
Ah, @martindholmes pointed out a good idea, we should test the schema I posted last week. So we did, and found an error (`rt/@target` was required, when it should be optional). Fixed in [this version](https://github.com/TEIC/TEI/files/5822122/issue_2054.zip), which also
a) constrains content of `<ruby>` to `rb, rt+`, and
b) has the file @martindholmes posted above as ruby.xml, with proper xml-model PIs for use of the schema.

**Comment by martindholmes** [comment]
We're now working in this temporary repo:
https://github.com/martindholmes/rubyForTEI

**Comment by martindholmes** [comment]
@sydb Our scheme doesn't yet answer @duncdrum's requirements above for `@hand` on r:rt. Do you agree with adding r:rt to att.written? I'm not sure what's implied by the more general request for "all the usual manuscribal attributes and elements". As far as I can see, it already has what it needs in terms of the global attributes and child content.

**Comment by sydb** [comment]
Oooh, good point. I suspect he meant att.transcriptional, which gets you `@status`, `@cause`, `@seq`, `@evidence`, `@instant`, and `@hand`. Does that set seem reasonable for `<rt>`? For `<rb>` and `<ruby>`? (The content models of `<rb>` and `<rt>` already allow for the various elements for manuscript transcription like `<redo>`.)

**Comment by martindholmes** [comment]
I'm really not sure. Since rb is part of the main transcription flow, I would think that if something applies to it, then you could wrap the entire ruby element, or if something applies to part of it, you would use a nested element. rt is different because it's outside the regular flow, and (presumably) is sometimes added by a different hand; as a new added unit, it would probably need all these attributes. I don't think the ruby element needs these things. @duncdrum What do you think?

**Comment by duncdrum** [comment]
`rt`s can be from different hands, and need the whole `att.transcriptional`, obviously in a rubicated manuscript `rb` needs  to be a member of that too, as multiple hands can be responsible for each. 

As for `ruby` It looks like a convenient shortcut in cases where base and target have been produced together (also very common). But I m not hellbent on its membership. Where would you put the attributes in question in such a case if not on ruby though, without duplicating base attributes on each target? Maybe a guideline rule that attributes on `rb` cover `rt` as well unless the latter contains an overriding attribute?

**Comment by martindholmes** [comment]
@duncdrum I'm not sure I understand why rb needs this stuff. It's part of the regular text flow, so you can wrap the entire ruby element in a hi element or a del element or any of those things if you want to apply some of these attributes to the whole of the rb; if you only want to apply them to part of the rb, then you simply tag that part inside the rb. rt is obviously different because it's not part of a continuous context, it's an atomic addition.

Although of course you could add the required tagging inside rt, so perhaps none of these elements needs all this external stuff.

**Comment by duncdrum** [comment]
> Does that set seem reasonable for `<rt>`? For `<rb>` and `<ruby>`?  

@martindholmes i was responding to the question for each. `rt` needs it, I m still unclear what the suggested encoding is for handwritten text, e.g. where to put  `@Hand` in cases where 

- Both base and target are written by the same hand, but there are multiple hands per document
- Base and targets are written by differing hands. 

Depending on that we can come up on guidelines prose and examples and class membership.

**Comment by martindholmes** [comment]
@duncdrum In either case, you could do:

`<rt><seg hand="#Fred">abc</seg></rt>`
`<rb><seg hand="#Fred">abc</seg></rb>`

That would only be necessary for rb on the assumption that the rb string is written by a different hand from the preceding or following text; otherwise, a higher-level element outside the `<ruby>` element would carry the `@hand`, no?

I think it would probably be convenient to allow this:

`<rt hand="#Fred">abc</rt>`

but I think the case where the rb and only the rb (not its preceding and following text) is in a distinct hand will be very unusual, don't you think?

**Comment by duncdrum** [comment]
Not at all. The imperial letter, like the one I showed in my presentation have min 1 scribe taking dictation, and signs from the emperor. Those are all base characters. Upon receiving the letter in e.g. Japan, there will Ruby annotations for the Japanese emperor by officials, and marks by the Japanese emperor signaling receipt, plus any further comments all potentially with rb/rt by other officials. 

Now imagine encoding a collection of imperial letters from the archives…

**Comment by martindholmes** [comment]
So it's very common in these texts to have a piece of text added in a fresh hand which consists only of a single rb, which is then glossed in a separate hand? I think I'd do this with:

~~~~~
<add hand="#Fred">
  <ruby>
    <rb>Thing</rb>
    <rt hand="#Mary">gloss</rt>
  </ruby>
</add>
~~~~~

Why would that not work?

**Comment by knagasaki** [comment]
> So it's very common in these texts to have a piece of text added in a fresh hand which consists only of a single rb, which is then glossed in a separate hand? I think I'd do this with:
> 
> ```
> <add hand="#Fred">
>   <ruby>
>     <rb>Thing</rb>
>     <rt hand="#Mary">gloss</rt>
>   </ruby>
> </add>
> ```

I think it is enough for the situation which @duncdrum described.
Otherwise, the following markup might be an alternative.

```xml
<signed hand="#Fred">
   <ruby>
     <rb>Thing</rb>
     <rt hand="#Mary">gloss</rt>
   </ruby>
</signed>
```
I hope @duncdrum would give some examples if both would not be appropriate.

> Why would that not work?

**Comment by duncdrum** [comment]
@martindholmes  that works great for such an addition to the main text and is what I would want to do in such a case. However, we keep discussing rubrication of of a single term, based on the gdocs example. What I'm trying to get at, is us thinking about how this will work in practice for a whole document. E.g. the main text shouldn't be wrapped in `<add>`. So let me escalate the snippet a bit. 

```xml
<body>
    <ab hand="#official_C_01">
    <ruby>
        <rb>This</rb>
        <rt hand="#official_J_03">Zis</rt>
        <rt hand="#official_J_02">Kindly</rt>
    </ruby>
    <ruby>
        <rb>be</rb>
        <rt hand="#official_J_02">regard this</rt>
    </ruby>  
    <ruby>
        <rb>text</rb>
        <rt hand="#official_J_03">tekst</rt>
        <rt hand="#official_J_02">letter</rt>
    </ruby>
    for you.
    </ab>
    <add hand="#official_J_01">
    <ruby>
        <rb>receiev</rb>
        <rt hand="#official_J_02">…ed</rt>
    </ruby>
    </add>
</body>
```

This work and is unambiguous without `@hand` on `<ruby>` or `<rb>`, still 
```xml
<ruby>
    <rb hand="#Fred">Thing</rb>
    <rt hand="#Mary">gloss</rt>
</ruby>
```

is more concise (one less element) and since we have to encode rubys for individual characters (or words) this might add up quickly. If we want to have a possibility to include whole sentences or larger segments in `<rb>` because the rubrication is standardized and regular, I think allowing the second option might make more sense. Basically:

```xml
<p style="writing-mode: vertical-rl" hand='#C1'> 
  <r:ruby>
    <r:rb>打<anchor/>球<anchor/>場</r:rb>
    <r:rt hand='#J1' place="left">ビリ<anchor/>アー<anchor/>ド</r:rt>
    <r:rt hand='#J2' place="right">ダ<anchor/>キュウ<anchor/></r:rt>
  </r:ruby> 
</p>
```

or for consistent examples (2 elements less than non-standoff above, and more readable imv):
```xml
<ab style="writing-mode: vertical-rl" hand='#C1'> 
  <r:ruby>
    <r:rb>This<anchor/>be<anchor/>text</r:rb>
    <r:rt hand='#J1' place="left">Zis<anchor/><anchor/>tekst</r:rt>
    <r:rt hand='#J2' place="right">Kindly<anchor/>regard this<anchor/>letter</r:rt>
  </r:ruby> 
  for you. 
</ab>
```

Lastly, with some note in the header:
```xml
<ab style="writing-mode: vertical-rl" hand='#C1'> 
  <r:ruby>
    <r:rb>This be text</r:rb>
    <r:rt hand='#J1' place="left">Zis <supplied reason="symmetry">bee</supplied> tekst</r:rt>
    <r:rt hand='#J2' place="right">Kindly <group type="ruby">regard this</group> letter</r:rt>
  </r:ruby> 
  for you. 
</ab>
```

maybe a `<rg>` ruby group element could help avoid `<anchor>` overload.

| markup | elementals count |
|---|---|
| non standoff | 13 |
| standoff w. anchor | 11 |
| standoff no anchor | 7 |

**Comment by duncdrum** [comment]
On a separate note: Can `<rt>` have `xml:lang` in the current proposal? It should, since the rubys can be different from the base text.

**Comment by knagasaki** [comment]
> On a separate not can `<rt>` have `xml:lang` in the current proposal? It should, since the rubys can be different from the base text.

I completely agree that `<rt>` element belongs to att.global! The proposal doesn't disagree to it, but may indicate it explicitly like below:
```xml
<p style="writing-mode: vertical-rl"> 
  <r:ruby>
    <r:rb hand='#C1'>打<anchor/>球<anchor/>場</r:rb>
    <r:rt hand='#J1' place="left" xml:lang="en">ビリ<anchor/>アー<anchor/>ド</r:rt>
    <r:rt hand='#J2' place="right">ダ<anchor/>キュウ<anchor/></r:rt>
  </r:ruby> 
</p>
```

**Comment by knagasaki** [comment]
> @martindholmes that works great for such an addition to the main text and is what I would want to do in such a case. However, we keep discussing rubrication of of a single term, based on the gdocs example. What I'm trying to get at, is us thinking about how this will work in practice for a whole document. E.g. the main text shouldn't be wrapped in `<add>`. So let me escalate the snippet a bit.
> 
> ```
> <body>
>     <ab @hand="#official_C_01">
>     <ruby>
>         <rb>This</rb>
>         <rt hand="#official_J_03">Zis</rt>
>         <rt hand="#official_J_02">Kindly</rt>
>     </ruby>
>     <ruby>
>         <rb>be</rb>
>         <rt hand="#official_J_02">regard this</rt>
>     </ruby>  
>     <ruby>
>         <rb>text</rb>
>         <rt hand="#official_J_03">tekst</rt>
>         <rt hand="#official_J_02">letter</rt>
>     </ruby>
>     for you.
>     </ab>
>     <add hand="#official_J_01">
>     <ruby>
>         <rb>receiev</rb>
>         <rt hand="#official_J_02">…ed</rt>
>     </ruby>
>     </add>
> </body>
> ```
> 
> This work and is unambiguous without `@hand` on `<ruby>` or `<rb>`, still
> 
> ```
> <ruby>
>     <rb hand="#Fred">Thing</rb>
>     <rt hand="#Mary">gloss</rt>
> </ruby>
> ```

I think that if another scribe writes a text, it should be encoded by any appropriate element rather than `<rb>` and @hand value should be inherited from the parent element. If #Fred added a word, it should be encoded by `<add>`. If signed, by `<signed>`...

**Comment by duncdrum** [comment]
@knagasaki yes i just noticed that there will always be a parent element that can take `@hand` and adjusted my examples. So we don't need `att.transcribed` membership for `ruby` or `rb`. 

i still like to move the discussion into the direction of considering more than annotations of single terms. As we need to avoid ambiguity and at least have an idea of what we want encoders to do, so we can write the guidelines prose and pick examples.

**Comment by kzhr** [comment]
Lubrication is not a targeted case for ``<ruby>``. I am not sure why it is related here.

As for the case where more than two ruby texts are given for a single body text, encoders should aware whether they encode parallel ruby texts, or a nested ruby text. It's just not about simplicity: it's about structure.
As a rule, the secondary ruby texts are added as additions to the primary ruby text. In a rare case, there can be a true parallel ruby texts; so, multiple ``rt``s may not be ruled out in every case. In any case, the preference should be put on nested ruby encoding.

**Comment by duncdrum** [comment]
@kzhr Yes nested rubys together with non-standoff markup are the only way to go when there are multiple complex ruby streams. 
```xml
<p style="writing-mode: vertical-rl" hand="#C1">
    <r:ruby>
        <r:rb>
            <r:ruby>
                <r:rb>打</r:rb>
                <r:rt place="right" hand='#J1'>ダ</r:rt>
            </r:ruby> 
            <r:ruby>
                <r:rb>球</r:rb>
                <r:rt place="right" hand='#J1'>キュウ</r:rt>
            </r:ruby>
            場
        </r:rb>
        <r:rt place="left" hand='#J2'>ビリアード</r:rt>
    </r:ruby>
</p>
```
Yet there are many instances of rubys in which there simply isn't a second stream at all. Think of bopomofo in school textbooks. The question is do we want to force encoders to adopt a character (or word) by character markup in such cases, when a more concise standoff approach would do as well?

**Comment by martindholmes** [comment]
Thank you everyone for this input. Yes, all elements get global attributes including `@xml:lang` by default, but it would be good to give an example of this. Can I put out a general call for interesting examples with good-quality images? We need images we can use in the Guidelines, so we need to have permission to use them from any copyright holder.

The billiards example is brilliant because it encapsulates so many  aspects of the problem in a single example, so I do want to keep that -- could we get a better image of it?

**Comment by kzhr** [comment]
> Yet there are many instances of rubys in which there simply isn't a second stream at all. Think of bopomofo in school textbooks. The question is do we want to force encoders to adopt a character (or word) by character markup in such cases, when a more concise standoff approach would do as well?

I am not sure whether standoff approach is concise for bopomofo texts. If encoders would like to do standoff markup (I'm not sure what is the benefit, though), they have an option of the pair of ``<seg>`` and ``<note>``. By using ruby environments, one will be benefited from clearer structure annotation. Excuse me if I missed the point.

**Comment by kzhr** [comment]
> Thank you everyone for this input. Yes, all elements get global attributes including `@xml:lang` by default, but it would be good to give an example of this. Can I put out a general call for interesting examples with good-quality images? We need images we can use in the Guidelines, so we need to have permission to use them from any copyright holder.
> 
> The billiards example is brilliant because it encapsulates so many aspects of the problem in a single example, so I do want to keep that -- could we get a better image of it?

As for billiards example, a finer and free image can be obtained from this URL: http://school.nijl.ac.jp/kindai/NIJL/NIJL-01116.html#37 (The right half, the 5th line)
![GetImage cgi](https://user-images.githubusercontent.com/964511/105491782-370b5100-5cfa-11eb-84b5-0877d7df6d12.jpg)

**Comment by duncdrum** [comment]
> If encoders would like to do standoff markup (I'm not sure what is the benefit, though), they have an option of the pair of `<seg>` and `<note>`. By using ruby environments, one will be benefited from clearer structure annotation. Excuse me if I missed the point.

`<note>` and `<seg>` are general purpose elements which can be used for encoding other textual features alongside rubys, users would be better off incorporating html5 elements as foreign namespace elements compared to these. 

> Such a textual structure requires another layer of semantics besides the existing set of vocabulary, such as <seg>, <note> 

If there is a specific element available, as a global standard TEI needs to consider its application in all kinds of CJK documents modern and historical, as well as different written languages. 

>  I am not sure whether standoff approach is concise for bopomofo texts. 

See my examples above, standoff is more concise for simple cases of rubication (irrespective of language), it can be combined with non stand-off where necessary (if at all).  E.g. Documents containing a single ruby stream, will never require nested `<ruby>` structures. But character-by-character markup will quickly interfere with other markup (linguistic, named entities, etc)  leading to overlap

**Comment by martindholmes** [comment]
Thank you @kzhr!

**Comment by duncdrum** [comment]
As for standoff there is of course the option to use `&ZeroWidthSpace;` `U+200B` which  `is intended for invisible word separation` instead of `<anchor/>` or any other xml elements to align segmentation of `<rb>` and `<rt>` i would at least like to mention that possibility in the guidelines section.

The advantages are better legibility as no visible word boundaries as in the CJK source, combined with the ability to process, index, using standard xpath string operations.

**Comment by martindholmes** [comment]
@duncdrum I think that would be nightmarish -- imagine trying to maintain a file which delimited content with zero-width spaces. I really don't think we should ever recommend something like that in the Guidelines when there are so many better alternatives.

**Comment by duncdrum** [comment]
@martindholmes i don't see the nightmare. It is the unicode character reserved for exactly the purpose of introducing delimiters for writing systems that don't use typographic word delimiters. Unicode aware software applications, such as xml processors, should play nice with it, and it can simplify things greatly, if the contents of `tei:body/text()` aren't a single undelimited string of length 200k+

**Comment by martindholmes** [comment]
@duncdrum I know what it's for, and it's great for what it's for, but I don't think it's a good option for creating stand-off markup. You can only point to it with XPointers, and if you're using XPointers to point to characters, you might as well point to the characters that are actually the target of the ruby text.

**Comment by duncdrum** [comment]
the idea isn't to point to the character, just as you're not pointing to the various `<anchor/>` in the examples in the OP. You construct arrays or maps to return the annotation for a given position in the sequence. 
Granted its a matter of taste, but i don't see the immediate disadvantage of `<rt>` with 3x `&ZeroWidthSpace;` over 3x `<anchor/>`. Both are stopgaps that depending on task will have to be processed.  

```xml
<p style="writing-mode: vertical-rl"> 
            <r:ruby>
              <r:rb>打&ZeroWidthSpace;球&ZeroWidthSpace;場</r:rb>
              <r:rt place="left">ビリ&ZeroWidthSpace;アー&ZeroWidthSpace;ド</r:rt>
              <r:rt place="right">ダ&ZeroWidthSpace;キュウ&ZeroWidthSpace;</r:rt>
            </r:ruby> 
          </p>
```

**Comment by martindholmes** [comment]
This particular method of "pointing" isn't one I'd recommend anyway; I think if you're using anchors you should use them properly, with ids and pointers. But of course there's nothing to stop anyone from inventing their own correspondence method based on contiguity for their own project.

**Comment by kzhr** [comment]
> > I am not sure whether standoff approach is concise for bopomofo texts.
> 
> See my examples above, standoff is more concise for simple cases of rubication (irrespective of language), it can be combined with non stand-off where necessary (if at all). E.g. Documents containing a single ruby stream, will never require nested `<ruby>` structures. But character-by-character markup will quickly interfere with other markup (linguistic, named entities, etc) leading to overlap

It would complicate the situation if we include something that are not similar in structure. I haven't understood the structural similarity between lubrication and ruby. As stated in our proposal document, ruby is not for all interlinear gloss examples. Rather, it specialises in their special cases.

I see the need for parallel encoding only with the bopomofo texts with Romanisation. Still, aren't they special cases? I no longer disagree with allowing more than two ``<rt>``s with single ``<rb>`` for the special cases anyway.

As for overlapping, I can hardly imagine the actual case: if you would encode them in character basis, what would be interfered?

**Comment by martindholmes** [comment]
@kzhr The billiard-hall example is the perfect case for one `<rb>` and two `<rt>`s. As far as stand-off is concerned, any of the many normal TEI standoff approaches can be used, as well as custom approaches such as the one @duncdrum describes using the zero-width space; but I agree with you that in most cases standoff will not be needed, and I think that in the chapter prose, we should provide a single recommended standoff approach, with one good example.

However complex your linguistic encoding, as long as your texts have ids on the relevant elements, then you can point to them for the purposes of standoff ruby (and you can of course insert `<c>` elements where required so that you can point directly to a single character).

**Comment by duncdrum** [comment]
> It would complicate the situation if we include something that are not similar in structure. I haven't understood the structural similarity between lubrication and ruby. As stated in our proposal document, ruby is not for all interlinear gloss examples. Rather, it specialises in their special cases.

Please take a look at the images that i just posted into martin's repo and tell me where I m trying to include things which are not similar in structure. All of these are `ruby`, in at least html5, css, and probably more standards. I m in full agreement that `ruby` are not just interlinear glosses which is why `<note> + <seg>` is a red herring solution, and we need another element. But yes East-Asian documents are more complicated (from the point of view of TEI) than `打球場`. At the same time, not all occurrences of `<ruby>` require the full force of the markup in this proposal. As with other parts of the *Guidelines* we need rules that can deal with the most complex of cases, as well as a more light weight option, to avoid unnecessary markup overkill.

**Comment by knagasaki** [comment]
> > It would complicate the situation if we include something that are not similar in structure. I haven't understood the structural similarity between lubrication and ruby. As stated in our proposal document, ruby is not for all interlinear gloss examples. Rather, it specialises in their special cases.
> 
> Please take a look at the images that i just posted into martin's repo and tell me where I m trying to include things which are not similar in structure. All of these are `ruby`, in at least html5, css, and probably more standards. I m in full agreement that `ruby` are not just interlinear glosses which is why `<note> + <seg>` is a red herring solution, and we need another element. But yes East-Asian documents are more complicated (from the point of view of TEI) than `打球場`. At the same time, not all occurrences of `<ruby>` require the full force of the markup in this proposal. As with other parts of the _Guidelines_ we need rules that can deal with the most complex of cases, as well as a more light weight option, to avoid unnecessary markup overkill.

Could you give me a link to or an URL of the image which you posted into Marti's repo?

**Comment by martindholmes** [comment]
@knagasaki These are Duncan's images:
https://github.com/martindholmes/rubyForTEI/tree/main/Images
I don't know which ones in particular are relevant to this point.

**Comment by kzhr** [comment]
> > It would complicate the situation if we include something that are not similar in structure. I haven't understood the structural similarity between lubrication and ruby. As stated in our proposal document, ruby is not for all interlinear gloss examples. Rather, it specialises in their special cases.
> 
> Please take a look at the images that i just posted into martin's repo and tell me where I m trying to include things which are not similar in structure. All of these are `ruby`, in at least html5, css, and probably more standards. I m in full agreement that `ruby` are not just interlinear glosses which is why `<note> + <seg>` is a red herring solution, and we need another element. But yes East-Asian documents are more complicated (from the point of view of TEI) than `打球場`. At the same time, not all occurrences of `<ruby>` require the full force of the markup in this proposal. As with other parts of the _Guidelines_ we need rules that can deal with the most complex of cases, as well as a more light weight option, to avoid unnecessary markup overkill.

Well, we intentionally excluded the most of manuscript examples, seeing that it is totally capable of existing `<note>` and `<seg>` (you can use `<anchor>` anyway) combination, because ruby is a special case where glossed text constitutes the text itself in parallel with the main line. For ease of encoding such glosses, we here propose the special elements. Putting all interlinear glosses are without doubt out of our scope, and I personally believe that it complicates TEI's semantic encodings.

**Comment by duncdrum** [comment]
@martindholmes @knagasaki check the commit messages, they describe the images and their context. martindholmes/rubyForTEI@ef301bd7324fa8879264f2fb0571d74c9b051eca is the manuscript example. 

@kzhr 
> Well, we intentionally excluded the most of manuscript examples, 

I thought we had already settled that `att.transcriptional` is necessary on `<rt>`? Using `<ruby>` to encode ruby only in print, but not in manuscripts makes no sense. 

>  Putting all interlinear glosses are without doubt out of our scope

agreed, i m still not sure why you think i m trying to do that. Can you tell me which examples are out of scope, and more importantly better describe as interlinear gloss instead of `<ruby>`? 

I m particularly curious about martindholmes/RubyForTEI@1b9cbe5fa5f57447a67dfb4c77676c21cc397599 on what basis is there interlinear glosses and not ruby? How should we encode interspersed emphasis markers with ruby annotations? How does this not consitute an special case for East Asian documents?

**Comment by knagasaki** [comment]
> @martindholmes @knagasaki check the commit messages, they describe the images and their context. [martindholmes/rubyForTEI@ef301bd](https://github.com/martindholmes/rubyForTEI/commit/ef301bd7324fa8879264f2fb0571d74c9b051eca) is the manuscript example.

The interlinear texts in the example pages can be encoded by &lt;ruby&gt; and &lt;note type="送り"&gt; and &lt;metamark function="kaeriten"&gt;. We use &lt;seg type="ruby"&gt; and rt, rb instead of &lt;ruby&gt; so far. It's usual case in Japanese manuscript. The first two lines can be done like below:
```xml
            <opener>
               <salute>上天
               <seg type="ruby">
                  <seg type="rb">眷</seg>
                  <seg place="right" type="rt">カエリ</seg>
               </seg>
               <note place="right" type="送り">ミル</note>
                  命
                  <note type="送り">ヲ</note>
               </salute>
                  大蒙古国
               <note place="right" type="送り">ノ</note>
               皇帝奉
               <note place="right" type="送り">ル</note>
               書</opener>
```

In some cases we add `@targetEnd` to &lt;note type="送り"&gt; with an anchor tag having the targeted xml:id by the `@targetEnd` in order to show the targeted range like below:

```xml
<note type="送り" targetEnd="#note01">ミル</note>
<seg type="ruby">
  <seg type="rb">眷</seg>
   <seg type="rt">カエリ</seg>
</seg>
<anchor xml:id="note01"/>
               
```
We are considering whether it should be an independent element or not.  Otherwise, some researchers might want to encode all of the interlinear texts by `<note type="訓">` instead of separating them into `<ruby>` and `<note type="送り">` in order to avoid complication of interpretation. The usage of the elements would depends on each project and community in the case of the pre-modern documents.
Moreover, we are considering to propose an element which represents a doubled line which is usually used as a gloss of a term. It might be able to be represented by `<note>` or `<gloss>` with an appropriate `@type`, but it might be sometimes difficult to be processed because it almost has one or more inner line breaks, as you know. As we basically want to follow the TEI P5 guidelines as far as possible, we will consider the possibility continuously.

**Comment by martindholmes** [comment]
@duncdrum @knagasaki @kzhr I think we all need to agree very clearly on the scope and intention of the ruby elements before we can write a useful prose section for the Guidelines on their use. I think we're all in agreement on the schema requirements now, but I sense that @duncdrum and @kzhr have slightly different views on where the use of these elements is appropriate and where it is not. 
As @knagasaki notes above, it is very common for TEI to provide multiple ways of encoding the same phenomenon; this is not ideal but it is to some extent inevitable. So in principle I have nothing against @duncdrum using ruby elements to encode something that @kzhr believes should be encoded in some other way. But I do believe that it is important to understand these different points of view so that we can explain them in the prose.

**Comment by kzhr** [comment]
@duncdrum 

> I thought we had already settled that `att.transcriptional` is necessary on `<rt>`? Using `<ruby>` to encode ruby only in print, but not in manuscripts makes no sense.

Agreed.

**Comment by duncdrum** [comment]
> Moreover, we are considering to propose an element which represents a doubled line which is usually used as a gloss of a term. It might be able to be represented by <note> or <gloss> with an appropriate @type, but it might be sometimes difficult to be processed because it almost has one or more inner line breaks, as you know.

@knagasaki yes commentarial insertions are certainly another case worthy to take a closer look at. From their structure I would actually say they more naturally resemble interlinear glosses in western codexes (as they don't have the kind of interaction with the base text flow the billiard hall example so nicely show). I'd be happy to take a look at means of encoding them with / without `@type` once the `ruby` things are approaching final form. 

Generally, i consider the target scope of `ruby` in TEI, to be broadly similar to that of `ruby` in e.g. `html5`. With the additional caveat that we need to consider historical usage, and full-length books/documents as application examples. So I'm still going through the list of my [original post](https://github.com/TEIC/TEI/issues/2054#issuecomment-720379521) in this thread:
-  We ve covered *manuscript examples in multiple hands* (although the copyright free image i found is a single hand copy of the original document) @martindholmes i think the copy will do for the *Guidelines*, if we add a sentence that it is a copy and that the original is in multiple hands?
- Next up, *emphasis markers interspersed with `ruby`*. see martindholmes/rubyForTEI#2.

To facilitate our discussion i have create a number of issues in martin's repo, so we can discuss them alongside the examples that are already in there (more are welcome). This thread is getting long.

**Comment by martindholmes** [comment]
I'm now completely confused about whether ruby is to be identified by its physical/typographical characteristics (which was my basic initial assumption) or by its semantic intent (which seems to be the basis for distinguishing it from interlinear glosses). If our definition of ruby has nothing to do with typography and everything to do with semantic intent, then before we can go any further we must enumerate all the specific usages which we can all agree _are_ ruby, along with representative examples of things which _look_ like ruby, but aren't.

**Comment by knagasaki** [comment]
> I'm now completely confused about whether ruby is to be identified by its physical/typographical characteristics (which was my basic initial assumption) or by its semantic intent (which seems to be the basis for distinguishing it from interlinear glosses). If our definition of ruby has nothing to do with typography and everything to do with semantic intent, then before we can go any further we must enumerate all the specific usages which we can all agree _are_ ruby, along with representative examples of things which _look_ like ruby, but aren't.

I think (and the original proposal indicated) that ruby is a type of interlinear glosses, which represents pronunciation of the targeted text and can be regarded as a parallel text.

**Comment by knagasaki** [comment]
> I'm now completely confused about whether ruby is to be identified by its physical/typographical characteristics (which was my basic initial assumption) or by its semantic intent (which seems to be the basis for distinguishing it from interlinear glosses). If our definition of ruby has nothing to do with typography and everything to do with semantic intent, then before we can go any further we must enumerate all the specific usages which we can all agree _are_ ruby, along with representative examples of things which _look_ like ruby, but aren't.

I think the examples which are described in the original proposal are enough for basic usage. While consideration of complex examples is fun and important for some users, I would like to brush up the proposed ones to put into the Guidelines, if possible. I would be glad if you would list what should we do for it.

**Comment by knagasaki** [comment]
> This particular method of "pointing" isn't one I'd recommend anyway; I think if you're using anchors you should use them properly, with ids and pointers. But of course there's nothing to stop anyone from inventing their own correspondence method based on contiguity for their own project.

I completely agree.

**Comment by martindholmes** [comment]
@knagasaki Thanks for the clarification. I'm checking in with Council to get their input, but my hope is that we can add the elements and a basic introduction into the next release of the Guidelines. My plan is to draft the chapter prose over the next few days, in the rubyForTei repository, and present it on Saturday.

**Comment by knagasaki** [comment]
> @knagasaki Thanks for the clarification. I'm checking in with Council to get their input, but my hope is that we can add the elements and a basic introduction into the next release of the Guidelines. My plan is to draft the chapter prose over the next few days, in the rubyForTei repository, and present it on Saturday.

Thank you very much for your intention. I will support it as far as I can.

**Comment by martindholmes** [comment]
There is a rough draft of the introductory section here:
https://hcmc.uvic.ca/~mholmes/rubyForTEI/ruby.html
Please let me know what you think. I know there are many more interesting variants and edge cases and annotation types to address, but this is intended to be enough to introduce the concept and some straightforward encoding examples that will handle 75% of cases.

**Comment by knagasaki** [comment]
> There is a rough draft of the introductory section here:
> https://hcmc.uvic.ca/~mholmes/rubyForTEI/ruby.html
> Please let me know what you think. I know there are many more interesting variants and edge cases and annotation types to address, but this is intended to be enough to introduce the concept and some straightforward encoding examples that will handle 75% of cases.

Thank you very much for the appropriate prose. The example of "TEI" is so funny.
Only thing that I would like to revise is that the ruby text "キュウ" is better to be "キウ". "キュウ" is exact writing today. But the example is based on an older writing system and the markup is almost literal.  And "ビリアード" is "ビリヤード" as well.

**Comment by martindholmes** [comment]
Thanks indeed! I'll fix that. Sorry for my poor transcription.

**Comment by martindholmes** [comment]
Fixed. I've also added some prose on the default place assumption (above for horizontal, right for vertical text).

**Comment by martindholmes** [comment]
This is a copy of the original posting by @martinascholger on the rubyForTEI repo; I'm reposting here because we need more Council members to respond. Please signal your preferences:

Poll on possible locations for the main text

| section in the GLs | full section* | reference** | comment
|---|---|---|---|
| Core: before or after [notes](https://tei-c.org/release/doc/tei-p5-doc/en/html/CO.html#CONO) | MDH, SB, RV, HC, MB, MS, JJ |  | MH thinks it will be a good fit with exactly this context, and it will be easy to segue into it. SB: _after_ notes; i.e. the new 3.8.2.  HBS thinks this is a very specific type of annotation and, as a user, she is not sure if she would look for it here, but as long as it is referenced in other modules, it works, of course. HC wonders whether it would be better in conjunction with `<gloss>` ([3.3.4](https://tei-c.org/release/doc/tei-p5-doc/en/html/CO.html#COHQU)), though perhaps that should be busted out of 3.3 "Highlighting and Quotation", where I'm not sure it really belongs anyway. Maybe a new 3.4 on "Glosses and Ruby"? ; MB likes pairing Glosses and Ruby; MS likes glosses +  ruby too; only concern are with the more advanced examples.  JJ: After 3.8.1, as a new 3.8.2 (with the current 3.8.2 and following being renumbered). It would rather overwhelm 3.3.4 as is. Like HC's idea of moving 3.3.4 elsewhere because it's not "Highlighting and Quotation" at all. Agree with idea of new 3.4 for "Glosses and Ruby". "term" belongs elsewhere; it's been put with "gloss" only because terms are sometimes glossed. 
| Linking / Segmentation / Alignment: between [16.11](https://tei-c.org/release/doc/tei-p5-doc/en/html/SA.html#SASOann) and [16.13](https://tei-c.org/release/doc/tei-p5-doc/en/html/SA.html#SAref) |  | MDH, SB, RV, HBS, HC, MS, JJ | (This means a new paragraph in 16.12, right? —SB) HC thinks if it didn't go in Core, it should go here. JJ says new short paragraph in 16.12 just to alert users where to look.|
| Simple Analytic Mechanisms: between [17.2](https://tei-c.org/release/doc/tei-p5-doc/en/html/AI.html#AIATTS) and [17.4](https://tei-c.org/release/doc/tei-p5-doc/en/html/AI.html#AILA)  |  | MDH [SB], HBS | *** |
| as part of [Writing Modes](https://tei-c.org/release/doc/tei-p5-doc/en/html/WD.html#WDWM) |  | MDH | When we come to deal with emphasis marks in combination with ruby, this might be the place for that. JJ says it would make more sense to link from the ruby section to the section on writing modes because ruby is often vertical; writing mode doesn't imply ruby annotations. |
| Primary Source: between [11.3](https://tei-c.org/release/doc/tei-p5-doc/en/html/PH.html#PHST) and [11.4](https://tei-c.org/release/doc/tei-p5-doc/en/html/PH.html#PHLAY) |HBS | MDH [SB] RV MS JJ | † |
| Primary Source: between [11.4.](https://tei-c.org/release/doc/tei-p5-doc/en/html/PH.html#PHLAY) and [11.5](https://tei-c.org/release/doc/tei-p5-doc/en/html/PH.html#PHSK) |  | MDH [SB] RV MS [JJ] | †  |

*the full section on ruby glosses should go there
**there should be a reference to ruby section
***SB suggests a reference in a new paragraph in 17.4 before start of 17.4.1
†SB leans towards a mention of ruby as part of both 11.3 and 11.4 in the paragraphs before the 1st subsection of each.   HBS agrees with this.

**Comment by martindholmes** [comment]
Thanks for the feedback on the location for the Ruby prose. The consensus is clear -- it should be in core -- and the comments suggest that the most appropriate approach will be to integrated it with the content relating to glosses, which is currently in 3.3.4, "Terms, Glosses, Equivalents, and Descriptions". That sub-subsection is already quite long, so what I propose to attempt is this:

Take 3.3.4 "Terms, Glosses, Equivalents, and Descriptions" and 3.3.5 "Some Further Examples" and make them into a new subsection (i.e. one level higher), which will become 3.4.

Title the new section "Terms, Glosses, Equivalents, Descriptions, and Ruby". (JJ suggests that terms belongs elsewhere, but that would be a major reorg, and I don't think we have time for that right now.)

Create subsections:

3.4.1 "Terms and Glosses" covering the first part of what is now 3.3.4. 
       This would also include the "Further Examples" bit, which is 
       really an extension of this.

3.4.2 "Ruby Annotations" containing the new prose already proposed.

3.4.3 "Equivalents and Descriptions" containing the part of what is now    
       3.3.4 which discusses altIdent, desc, and equiv; I think this is 
       a different topic, really, from glossing terms.

I'm doing this work in a separate branch:

https://github.com/TEIC/TEI/tree/issue-2054-ruby

which I'll merge dev into steadily as I go, to minimize later merge conflicts.
